### DISCUSSION TITLE
How to Scan a QR code?

### DISCUSSION THREAD

**Discussion by pirasanthan-jesugeevegan** [appium/appium#17582] [Q&A]
Appium & WebdriverIo (Javascript)

Hi Guys,

I have an app that i'm automating that requires to scan a QR code using its camera.
I was wondering if there is a way of scanning a Static png file
Or any other way to mock this scenario?

Note - There is no option to select image from device

**Comment by RSM-SBA** [comment]
i have found only a hacky way for android to inject an image to the virtual 3D camera scene
You can add the following lines into the file Torent1BD.posters within the emulator directory

poster custom
size 0.5 0.5
position 0 0 -1.8
rotation 0 0 0
default poster.png

This will put the file poster.png which has to to be in the same folder, within the 3d camera scene.
![image](https://user-images.githubusercontent.com/90029825/194525910-895bd967-0c1b-45ca-88f8-1ec400f21083.png)

The emulator has to be started with the avd commandline option "-camera-back virtualscene"

**Reply by pirasanthan-jesugeevegan** [reply]
This is great, Now I got to the point to add the QR code during the test. 

Like Executing Shell comand during test or some other way of adding the `poster.png` during test

**Reply by sathya147** [reply]
Thank you Stephan for the detailed steps.  I was able to place my intended QR Code image inside the emulator.  And upon clicking the QR code scanning button from my app, I'm able to successfully scan the QR code.  

The next thing I'm needing is a way to replace the image during code execution for the next test case.  
I'm working with **Webdriverio and Typescript.**  Help on this would be much appreciated.

**Comment by RSM-SBA** [comment]
i don't know how to write this in javascript but it should be similar

Java :
final String androidHome = System.getenv("ANDROID_HOME");
File source = new File("PATH_TO_YOUR_FILE_SRC_DIR"+"FILENAME");
File dest = new File(androidHome+"/emulator/resources/poster.png");
FileUtils.copyFile(source, dest);

But i noticed that you can not change the picture unlimited times while the emulator runs. The 3D scene is not reloaded everytime a new picture is copied to the folder. Maybe killing app and restarting it within the running test helps to reset the camera scene.

**Reply by RSM-SBA** [reply]
but if your main test lies behind the QR code and its not the main goal to test the QR code scanning itself, you should ask your developers to create a simple way to come behind the qr code. That's what we doing in iOS.
